client/context/broadcast.go

```diff
@@ -29,6 +34,37 @@ func (ctx CLIContext) BroadcastTx(txBytes []byte) (res sdk.TxResponse, err error
 	return res, err
 }
 
+// CheckTendermintError checks if the error returned from BroadcastTx
+// is a Tendermint error that is returned before the tx is submitted
+// due to precondition checks that failed. If an Tendermint error is detected,
+// this function returns the correct code back in TxResponse.
+func CheckTendermintError(err error, txBytes []byte) *sdk.TxResponse {
```

Comment:
Sighhh, so it seems we don't get the actual errors from Tendermint; The context is lost once it passes through the HTTP/RPC logic:

```go
func RPCInternalError(id jsonrpcid, err error) RPCResponse {
	return NewRPCErrorResponse(id, -32603, "Internal error", err.Error()) // we no longer have context here...
}
```

I'm afraid our only solution is to string match via `strings.Contains(err.Error(), "...")`.

Reply:
cc @melekes it would be nice if we didn't have to (partial) string match and instead `RPCResponse` have the ability to return the original (wrapped) error.

Reply:
Don't see a reason why we can't do that

Reply:
from json-rpc spec:
```
data:

A Primitive or Structured value that contains additional information about the error.
This may be omitted.
The value of this member is defined by the Server (e.g. detailed error information, nested errors etc.).
```

Reply:
@melekes not sure what you mean by `data`? We'd like the `RPCError` to not lose context of the original error so that we can properly type check/error patch.

Reply:
Looks like we are using msg to categorize the error type (Internal vs Method not found) but data is already used to store the err.Error() message. 
I think the only way to reuse data is to shift msg to store err.Error(), which means we need to wrap the error type there, and then store the err in Data.
I don't know who's the consumer of the RPCResponse and how they treat Message and Data here.

Reply:
hmmm not sure I follow, but I don't think we want overcomplicate this. What I'm proposing is that Tendermint simply adds a field `err error` to `RPCError` or provide some other way to recover the _original_ `error` so we can type match via Go's APIs. Right now, we can't do this.

Regardless of the actual solution and for the purposes of this PR we have to string match:

```go
func CheckTendermintError(err error, txBytes []byte) *sdk.TxResponse {
	switch {
	case strings.Contains(err.Error(), mempool.ErrTxInCache.Error()):
		return &sdk.TxResponse{
			Code:   uint32(sdk.CodeTxInMempoolCache),
			TxHash: string(tmhash.Sum(txBytes)),
		}

	case strings.Contains(err.Error(), "mempool is full"):
		return &sdk.TxResponse{
			Code:   uint32(sdk.CodeMempoolIsFull),
			TxHash: string(tmhash.Sum(txBytes)),
		}

	case strings.Contains(err.Error(), "Tx too large"):
		return &sdk.TxResponse{
			Code:   uint32(sdk.CodeTxTooLarge),
			TxHash: string(tmhash.Sum(txBytes)),
		}

	default:
		return nil
	}
}
```

This solution is brittle but I don't see any other way for the time being. Do you @melekes ?

Reply:
Ah I see, seems like this is the only thing we can do for now.

Reply:
Unforunately :-/